File: ios/ElderlyAssistant/Services/MedicationScheduler/ConfirmationChallenge.swift
import Foundation

// MARK: - Confirmation Challenge (FR-D01, FR-D02)

enum ConfirmationState: Equatable {
    case idle
    case awaitingResponse(deadline: Date)
    case confirmed
    case denied
    case timedOut
}

enum ConfirmationAction: Equatable {
    case issueChallenge(prompt: String)
    case recordTaken
    case reEnterEscalation
    case noAction
}

final class ConfirmationChallenge {
    let challengeTimeoutSeconds: TimeInterval   // default: 30 per FR-D02
    private(set) var state: ConfirmationState

    init(challengeTimeoutSeconds: TimeInterval = 30) {
        self.challengeTimeoutSeconds = challengeTimeoutSeconds
        self.state = .idle
    }

    func start(medicationName: String, confirmationDescription: String?) -> ConfirmationAction {
        let prompt: String
        if let desc = confirmationDescription, !desc.isEmpty {
            prompt = "You said you've taken your \(medicationName). Is that the \(desc)?"
        } else {
            prompt = "Did you take your \(medicationName) just now?"
        }
        let deadline = Date().addingTimeInterval(challengeTimeoutSeconds)
        state = .awaitingResponse(deadline: deadline)
        return .issueChallenge(prompt: prompt)
    }

    func userResponds(with response: ConfirmationResponse) -> ConfirmationAction {
        guard case .awaitingResponse(let deadline) = state else {
            return .noAction
        }

        if Date() > deadline {
            state = .timedOut
            return .reEnterEscalation
        }

        switch response {
        case .yes:
            state = .confirmed
            return .recordTaken
        case .no:
            state = .denied
            return .reEnterEscalation
        }
    }

    func timeoutExpired() -> ConfirmationAction {
        guard case .awaitingResponse = state else {
            return .noAction
        }
        state = .timedOut
        return .reEnterEscalation
    }

    func reset() {
        state = .idle
    }
}

enum ConfirmationResponse {
    case yes
    case no
}

File: android/app/src/main/java/com/elderlyassistant/services/medication/ConfirmationChallenge.kt
package com.elderlyassistant.services.medication

// MARK: - Confirmation Challenge (FR-D01, FR-D02)

sealed class ConfirmationState {
    object Idle : ConfirmationState()
    data class AwaitingResponse(val deadline: Long) : ConfirmationState()
    object Confirmed : ConfirmationState()
    object Denied : ConfirmationState()
    object TimedOut : ConfirmationState()
}

sealed class ConfirmationAction {
    data class IssueChallenge(val prompt: String) : ConfirmationAction()
    object RecordTaken : ConfirmationAction()
    object ReEnterEscalation : ConfirmationAction()
    object NoAction : ConfirmationAction()
}

enum class ConfirmationResponse {
    YES, NO
}

class ConfirmationChallenge(
    private val challengeTimeoutSeconds: Long = 30
) {
    var state: ConfirmationState = ConfirmationState.Idle
        private set

    fun start(medicationName: String, confirmationDescription: String?): ConfirmationAction {
        val prompt = if (!confirmationDescription.isNullOrEmpty()) {
            "You said you've taken your $medicationName. Is that the $confirmationDescription?"
        } else {
            "Did you take your $medicationName just now?"
        }
        val deadline = System.currentTimeMillis() + (challengeTimeoutSeconds * 1000)
        state = ConfirmationState.AwaitingResponse(deadline)
        return ConfirmationAction.IssueChallenge(prompt)
    }

    fun userResponds(response: ConfirmationResponse): ConfirmationAction {
        val awaiting = state as? ConfirmationState.AwaitingResponse
            ?: return ConfirmationAction.NoAction

        if (System.currentTimeMillis() > awaiting.deadline) {
            state = ConfirmationState.TimedOut
            return ConfirmationAction.ReEnterEscalation
        }

        return when (response) {
            ConfirmationResponse.YES -> {
                state = ConfirmationState.Confirmed
                ConfirmationAction.RecordTaken
            }
            ConfirmationResponse.NO -> {
                state = ConfirmationState.Denied
                ConfirmationAction.ReEnterEscalation
            }
        }
    }

    fun timeoutExpired(): ConfirmationAction {
        if (state !is ConfirmationState.AwaitingResponse) {
            return ConfirmationAction.NoAction
        }
        state = ConfirmationState.TimedOut
        return ConfirmationAction.ReEnterEscalation
    }

    fun reset() {
        state = ConfirmationState.Idle
    }
}
